### PR TITLE
Add Support for TOSHIBA TMPM066 board

### PR DESCRIPTION
**Pull request by ganesh-ramachandran**
Added Interface firmware for TMPM066 and Bootloader firmware for TMPM366

project-generator PR: project-generator/project_generator_definitions#122

### CODE REVIEW

**Comment by theotherjimmy** [comment]
Hey @ganesh-ramachandran It looks like you did a merge. We can't backport merge commits into patch releases, and we like to add new board support into patch releases. Could you do a rebase instead?

**Comment by ganesh-ramachandran** [comment]
@theotherjimmy since this page supported online merge, merge commit was done. Rebased now. Thanks.

File: test/info.py [hunk 1]
```diff
@@ -144,6 +145,8 @@
     (   0x1237,     'sam3u2c_ublox_evk_nina_b1_if',            'sam3u2c_bl',        'U-BLOX-EVK-NINA-B1'                    ),
     (   0xC006,     'lpc11u35_vbluno51_if',                     None,               'VNG-VBLUno51'                          ),
     (   0xC005,     'lpc11u35_mtconnect04s_if',                 None,               'MtConnect04S'                          ),
+    (   0xC006,     'lpc11u35_vbluno51_if',                     None,               'VNG-VBLUno51'                          ),
```
Comment (c1728p9) [inline test/info.py]: This board isn't using the tmpm366. Any reason it needs to be added in this PR? Also did you want to include it in the release table above?

File: source/hic_hal/toshiba/tmpm366/driver/inc/tmpm366_fc.h [hunk 2]
```diff
@@ -0,0 +1,127 @@
+/**
+ *******************************************************************************
+ * @file    tmpm366_fc.h
+ * @brief   This file provides all the functions prototypes for FC driver.
+ * @version V2.0.2.4
+ * @date    2014/09/28
+ * 
+ * (C)Copyright TOSHIBA ELECTRONIC DEVICES & STORAGE CORPORATION 2017 All rights reserved
+ *******************************************************************************
```
Comment (c1728p9) [inline source/hic_hal/toshiba/tmpm366/driver/inc/tmpm366_fc.h]: This and other files should have a permissive open source license.

Comment (ganesh-ramachandran) [reply source/hic_hal/toshiba/tmpm366/driver/inc/tmpm366_fc.h]: Corrected the copyright messages.

File: test/info.py (hunk 1 above)
Comment (ganesh-ramachandran) [reply test/info.py]: Removed the duplicate configuration

**Comment by ganesh-ramachandran** [comment]
@c1728p9 Rebased done but could not execute the test. Below is the log when we try to execute the test. Please suggest what we are missing.

```
$ python run_test.py --user ganeshramachandran --password ******** --firmware k20dx_frdmkl02z_if
SubTest: Building target KL25Z
Starting remote build
Traceback (most recent call last):
  File "run_test.py", line 701, in <module>
    main()
  File "run_test.py", line 587, in main
    build_target_bundle(target_dir, args.user, args.password, test_info)
  File "C:\Users\test\Desktop\DAPLink\test\target.py", line 91, in build_target_bundle
    directory)
  File "C:\Users\test\Desktop\DAPLink\test\mbedapi.py", line 81, in build_repo
    raise Exception("Error while talking to the mbed API")
Exception: Error while talking to the mbed API
```

**Comment by c1728p9** [comment]
Hi @ganesh-ramachandran, since the target images can't be built try running tests on just the new HIC. This can be done by running `python run_test.py --notestendpt --testdl --loadbl`.

**Comment by ganesh-ramachandran** [comment]
Hello @c1728p9,  Thank you for information, Below is the log when we execute 
`python run_test.py --notestendpt --testdl --loadbl`

```
$ python run_test.py --notestendpt --testdl --loadbl
Switching to APP mode on board: 000000000000000000000000000000000000000097969904
changing mode BL -> IF
unmount took 0.651000022888 s
mount took 3.16499996185 s
SubTest: Build test configuration
No board to test firmware tmpm366_tmpm066_stk_if
Nothing that can be tested
```

**Comment by c1728p9** [comment]
Hi @ganesh-ramachandran, you'll need to have the TOSHIBA TMPM066 board connected when you run this command. You could check to see if the board is properly detected by running `mbedls`. You should see something like this but with your board:
```
> mbedls
+---------------+----------------------+-------------+-------------+--------------------------------------------------+-----------------+
| platform_name | platform_name_unique | mount_point | serial_port | target_id                                        | daplink_version |
+---------------+----------------------+-------------+-------------+--------------------------------------------------+-----------------+
| K64F          | K64F[0]              | D:          | COM5        | 0240000033514e45001c500585d40025e981000097969900 | 0244            |
+---------------+----------------------+-------------+-------------+--------------------------------------------------+-----------------+
```

**Comment by ganesh-ramachandran** [comment]
Hello @c1728p9 , Thank you! We already connected the board but again we tried with a connected TMPM066 board and we tested in both mode(DAPLink & Maintenance) of interface chip, below is the log when we executed 
> python test/run_test.py --notestendpt --testdl --loadbl

```
mbedls
+---------------+----------------------+-------------+-------------+--------------------------------------------------+-----------------+
| platform_name | platform_name_unique | mount_point | serial_port | target_id                                        | daplink_version |
+---------------+----------------------+-------------+-------------+--------------------------------------------------+-----------------+
| TMPM066       | TMPM066[0]           | F:          | COM7        | 701100000000000000000000000000000000000097969904 | 0244            |
+---------------+----------------------+-------------+-------------+--------------------------------------------------+-----------------+
```

Interface chip in DAPLink mode
```
$ python test/run_test.py --notestendpt --testdl --loadbl
SubTest: Build test configuration
No board to test firmware tmpm366_tmpm066_stk_if
Nothing that can be tested
```

Interface chip in Maintenance mode
```
$ python test/run_test.py --notestendpt --testdl --loadbl
Switching to APP mode on board: 000000000000000000000000000000000000000097969904
changing mode BL -> IF
unmount took 0.642999887466 s
mount took 3.26600003242 s
SubTest: Build test configuration
No board to test firmware tmpm366_tmpm066_stk_if
Nothing that can be tested
```

**Comment by ganesh-ramachandran** [comment]
Hi @c1728p9, we have rebased again.

File: test/info.py [hunk 3]
```diff
@@ -78,7 +78,11 @@
     ("sam3u2c_ublox_evk_nina_b1_if",                True,       0x5000,     "bin"       ),
     ("lpc11u35_vbluno51_if",                        False,      0x0000,     "bin"       ),
     ('lpc11u35_mtconnect04s_if',                    False,      0x0000,     "bin"       ),
+<<<<<<< HEAD
     ('lpc11u35_mbed_cloud_connect_if',              False,      0x0000,     "bin"       ),
+=======
+    ('tmpm366_tmpm066_stk_if',                      None,       0x10000,    "bin"       ),
+>>>>>>> Add Support for Toshiba TMPM066/TMPM366
```
Comment (c1728p9) [inline test/info.py]: Looks like you have a merge conflict here. That may be why you can't run testing.

**Comment by ganesh-ramachandran** [comment]
Hi @c1728p9, Resolved the conflict, rebased and tested. We got the same test log https://github.com/ARMmbed/DAPLink/pull/295#issuecomment-338375745

**Comment by c1728p9** [comment]
Hi @ganesh-ramachandran 
From mbed-ls the board ID is shown as "701100000000000000000000000000000000000097969904" which indicates that the device is in interface mode. When you are running testing the board ID is "000000000000000000000000000000000000000097969904" which indicates the board is in bootloader mode. Is something switching your board into bootloader mode? Is it getting stuck in bootloader mode?

File: source/daplink/daplink.sct [hunk 4]
```diff
@@ -34,6 +34,9 @@ LR_IROM1 DAPLINK_ROM_APP_START DAPLINK_ROM_APP_SIZE {
    .ANY (ram_func)
    .ANY (+RW +ZI)
    .ANY (RAM1)
+#if defined(M366)
+   tmpm366_fc.o(+RO)
+#endif
```
Comment (c1728p9) [inline source/daplink/daplink.sct]: This linker script should not be customized. Why is this needed?

**Comment by ganesh-ramachandran** [comment]
Hi @c1728p9 , thank you for the inputs. We have corrected the .sct file and updated the board id. We were able to run the test, but few fails were observed. The test results log files are attached. Please suggest.

[requirements.txt](https://github.com/ARMmbed/DAPLink/files/1449587/requirements.txt)
[summary.txt](https://github.com/ARMmbed/DAPLink/files/1449588/summary.txt)
[tmpm366_tmpm066_stk_if AdBun-M066.txt](https://github.com/ARMmbed/DAPLink/files/1449589/tmpm366_tmpm066_stk_if.AdBun-M066.txt)

**Comment by ganesh-ramachandran** [comment]
Hi @c1728p9, Resolved the conflict, rebased and tested. 

```
$ python test/run_test.py --notestendpt
[1533809606.966000] SubTest: DAPLink
[1533809607.087000] SubTest: Build test configuration
[1533809607.088000] Test configurations to be run:
[1533809607.088000]     0: APP=tmpm366_tmpm066_stk_if BL=tmpm366_bl Board=AdBun-M066 Target=<None>
[1533809607.088000]
[1533809607.088000] All firmware can be tested
[1533809607.088000]
[1533809607.088000] SubTest: tmpm366_tmpm066_stk_if AdBun-M066
[1533809607.088000] Board: Name=AdBun-M066 Unique ID=701100000000000000000000000000000000000097969904
[1533809607.088000] Application: Name=tmpm366_tmpm066_stk_if
[1533809607.088000] Bootloader: Name=tmpm366_bl
[1533809607.088000] Target: None
[1533809607.088000] SubTest: load_interface
[1533809607.088000] SubTest: set_mode
[1533809607.088000] changing mode IF -> BL
[1533809607.096000] SubTest: wait_for_remount
[1533809607.935000] unmount took 0.838999986649 s
[1533809611.809000] mount took 3.87400007248 s
[1533809624.714000] programming took 12.3849999905 s
[1533809624.714000] SubTest: wait_for_remount
[1533809625.461000] unmount took 0.746999979019 s
[1533809628.652000] mount took 3.19099998474 s
[1533809628.652000] SubTest: set_mode
[1533809628.652000] Interface crc: 0x76995561

Test: tmpm366_tmpm066_stk_if AdBun-M066: Pass
No handlers could be found for logger "mbedls.lstools_base"
Traceback (most recent call last):
  File "test/run_test.py", line 701, in <module>
    main()
  File "test/run_test.py", line 685, in main
    local_changes=local_changes)
  File "test/run_test.py", line 279, in write_test_results
    stderr=subprocess.STDOUT)
  File "C:\yotta\python\lib\subprocess.py", line 186, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['pip', 'freeze']' returned non-zero exit status 2
```

Test output Pass (**Test: tmpm366_tmpm066_stk_if AdBun-M066: Pass**) followed by No handlers could be found for logger "mbedls.lstools_base". Please confirm.

**Comment by flit** [comment]
What OS did you run your tests on? The "No handlers" error can probably be safely ignored.

Any particular reason you tested with `--notestendpt`? Disabling the endpoint tests removes the majority of test cases.

**Comment by ganesh-ramachandran** [comment]
Hi @flit,  thank you for the inputs.

> What OS did you run your tests on? The "No handlers" error can probably be safely ignored.

We are using Windows 10.

> Any particular reason you tested with --notestendpt? Disabling the endpoint tests removes the majority of test cases.

--notestendpt is suggested by @c1728p9 : https://github.com/ARMmbed/DAPLink/pull/295#issuecomment-337595145

**Comment by ganesh-ramachandran** [comment]
Hi @flit @c1728p9, please confirm if any other update is required.

**Comment by brianesquilona** [comment]
Hi @ganesh-ramachandran, since the board is connected to your system, I will advise to run the whole test with the bootloader loaded --loadbl and interface test --testdl. 

As @flit mentioned the endpoints tests are also important. But you will need a daplink validation code, which is a simple binary, to test the serial, hid, msc. You will need to add this target binary somewhere and point the `--targetdir target_files` for example if you put it in the directory `target_files` at the root of the daplink folder. Your target binary should be named `AdBun-M066.bin` form the info.py. If you do not have this currently you need to add --notestendpt
python test\run_test.py --force --firmware tmpm366_tmpm066_stk_if --firmwaredir projectfiles\uvision --testdl --loadbl --notestendpt 

But the full command for the test with end pts is 
python test\run_test.py --force --firmware tmpm366_tmpm066_stk_if --targetdir target_files --firmwaredir projectfiles\uvision --testdl --loadbl.

--firmwaredir points to your build location

**Comment by flit** [comment]
Any further progress on this? If not, we will close this PR soon.

**Comment by flit** [comment]
Sorry! Accidentally closed too soon. 😄

**Comment by 0xc0170** [comment]
@ganesh-ramachandran What shall we do with this PR? It's based on quite old codebase, it needs rebase and update ? It might be easier to branch off master as it is now to add this target and also attach test results.

**Comment by ganesh-ramachandran** [comment]
Will revise and create new PR.
